From: Mark Hindley <mark@hindley.org.uk>
Date: Tue, 13 Dec 2022 16:04:06 +0000
Subject: [PATCH] Support libseat.

Devuan patch. Equivalent to support-libseat-upstream fc24510f17e89a5bbac1065abab758a4d0c42634


Closes: #799
---
 configure.ac                                |  18 ++
 hw/xfree86/common/xf86Events.c              |  24 +-
 hw/xfree86/common/xf86Init.c                |   3 +
 hw/xfree86/common/xf86Xinput.c              |   8 +-
 hw/xfree86/os-support/linux/Makefile.am     |   5 +
 hw/xfree86/os-support/linux/lnx_init.c      |   8 +-
 hw/xfree86/os-support/linux/lnx_platform.c  |  20 +-
 hw/xfree86/os-support/linux/seatd-libseat.c | 394 ++++++++++++++++++++++++++++
 hw/xfree86/os-support/shared/VTsw_usl.c     |   6 +
 include/Makefile.am                         |   1 +
 include/dix-config.h.in                     |   3 +
 include/meson.build                         |   1 +
 include/seatd-libseat.h                     |  49 ++++
 meson.build                                 |   8 +-
 meson_options.txt                           |   2 +
 15 files changed, 534 insertions(+), 16 deletions(-)
 create mode 100644 hw/xfree86/os-support/linux/seatd-libseat.c
 create mode 100644 include/seatd-libseat.h

diff --git a/configure.ac b/configure.ac
index f8145e4..1b5ad9f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -574,6 +574,7 @@ AC_ARG_ENABLE(pciaccess, AS_HELP_STRING([--enable-pciaccess], [Build Xorg with p
 AC_ARG_ENABLE(linux_acpi, AS_HELP_STRING([--disable-linux-acpi], [Disable building ACPI support on Linux (if available).]), [enable_linux_acpi=$enableval], [enable_linux_acpi=yes])
 AC_ARG_ENABLE(linux_apm, AS_HELP_STRING([--disable-linux-apm], [Disable building APM support on Linux (if available).]), [enable_linux_apm=$enableval], [enable_linux_apm=yes])
 AC_ARG_ENABLE(systemd-logind, AS_HELP_STRING([--enable-systemd-logind], [Build systemd-logind support (default: auto)]), [SYSTEMD_LOGIND=$enableval], [SYSTEMD_LOGIND=auto])
+AC_ARG_ENABLE(seatd-libseat, AS_HELP_STRING([--enable-seatd-libseat], [Build seatd-libseat support (default: auto)]), [SEATD_LIBSEAT=$enableval], [SEATD_LIBSEAT=auto])
 AC_ARG_ENABLE(suid-wrapper, AS_HELP_STRING([--enable-suid-wrapper], [Build suid-root wrapper for legacy driver support on rootless xserver systems (default: no)]), [SUID_WRAPPER=$enableval], [SUID_WRAPPER=no])
 
 dnl DDXes.
@@ -890,6 +891,23 @@ if test "x$CONFIG_HAL" = xyes; then
 fi
 AM_CONDITIONAL(CONFIG_HAL, [test "x$CONFIG_HAL" = xyes])
 
+if test "x$SEATD_LIBSEAT" = xauto; then
+        if test "x$CONFIG_UDEV" = xyes ; then
+                SEATD_LIBSEAT=yes
+        else
+                SEATD_LIBSEAT=no
+        fi
+fi
+if test "x$SEATD_LIBSEAT" = xyes; then
+        if ! test "x$CONFIG_UDEV" = xyes ; then
+                AC_MSG_ERROR([seatd-libseat is only supported in combination with udev configuration.])
+        fi
+        REQUIRED_LIBS="$REQUIRED_LIBS libseat"
+        AC_DEFINE(SEATD_LIBSEAT, 1, [Enable libseat integration])
+        SYSTEMD_LOGIND=no
+fi
+AM_CONDITIONAL(SEATD_LIBSEAT, [test "x$SEATD_LIBSEAT" = xyes])
+
 if test "x$SYSTEMD_LOGIND" = xauto; then
         if test "x$HAVE_DBUS" = xyes -a "x$CONFIG_UDEV" = xyes ; then
                 SYSTEMD_LOGIND=yes
diff --git a/hw/xfree86/common/xf86Events.c b/hw/xfree86/common/xf86Events.c
index 395bbc7..51a29c1 100644
--- a/hw/xfree86/common/xf86Events.c
+++ b/hw/xfree86/common/xf86Events.c
@@ -85,6 +85,7 @@
 
 #include "xf86platformBus.h"
 #include "systemd-logind.h"
+#include "seatd-libseat.h"
 
 extern void (*xf86OSPMClose) (void);
 
@@ -176,7 +177,9 @@ xf86ProcessActionEvent(ActionEvent action, void *arg)
             int vtno = *((int *) arg);
 
             if (vtno != xf86Info.vtno) {
-                if (!xf86VTActivate(vtno)) {
+                if (seatd_libseat_controls_session()) {
+                    seatd_libseat_switch_session(vtno);
+                } else if (!xf86VTActivate(vtno)) {
                     ErrorF("Failed to switch from vt%02d to vt%02d: %s\n",
                            xf86Info.vtno, vtno, strerror(errno));
                 }
@@ -185,7 +188,9 @@ xf86ProcessActionEvent(ActionEvent action, void *arg)
         break;
     case ACTION_SWITCHSCREEN_NEXT:
         if (!xf86Info.dontVTSwitch) {
-            if (!xf86VTActivate(xf86Info.vtno + 1)) {
+            if (seatd_libseat_controls_session()) {
+                seatd_libseat_switch_session(xf86Info.vtno + 1);
+            } else if (!xf86VTActivate(xf86Info.vtno + 1)) {
                 /* If first try failed, assume this is the last VT and
                  * try wrapping around to the first vt.
                  */
@@ -198,7 +203,9 @@ xf86ProcessActionEvent(ActionEvent action, void *arg)
         break;
     case ACTION_SWITCHSCREEN_PREV:
         if (!xf86Info.dontVTSwitch && xf86Info.vtno > 0) {
-            if (!xf86VTActivate(xf86Info.vtno - 1)) {
+            if (seatd_libseat_controls_session()) {
+                seatd_libseat_switch_session(xf86Info.vtno - 1);
+            } else if (!xf86VTActivate(xf86Info.vtno - 1)) {
                 /* Don't know what the maximum VT is, so can't wrap around */
                 ErrorF("Failed to switch from vt%02d to previous vt: %s\n",
                        xf86Info.vtno, strerror(errno));
@@ -219,7 +226,8 @@ xf86ProcessActionEvent(ActionEvent action, void *arg)
 void
 xf86Wakeup(void *blockData, int err)
 {
-    if (xf86VTSwitchPending())
+    if (xf86VTSwitchPending() ||
+        (dispatchException & DE_TERMINATE))
         xf86VTSwitch();
 }
 
@@ -309,6 +317,7 @@ xf86DisableInputDeviceForVTSwitch(InputInfoPtr pInfo)
 
     xf86ReleaseKeys(pInfo->dev);
     ProcessInputEvents();
+    seatd_libseat_close_device(pInfo);
     DisableDevice(pInfo->dev, TRUE);
 }
 
@@ -470,9 +479,10 @@ xf86VTEnter(void)
     dixSaveScreens(serverClient, SCREEN_SAVER_FORCER, ScreenSaverReset);
 
     for (pInfo = xf86InputDevs; pInfo; pInfo = pInfo->next) {
-        /* Devices with server managed fds get enabled on logind resume */
-        if (!(pInfo->flags & XI86_SERVER_FD))
+        /* Devices with server managed fds get enabled on logind/libseat resume */
+        if (!(pInfo->flags & XI86_SERVER_FD)) {
             xf86EnableInputDeviceForVTSwitch(pInfo);
+        }
     }
 
     for (ih = InputHandlers; ih; ih = ih->next) {
@@ -499,6 +509,8 @@ static void
 xf86VTSwitch(void)
 {
     DebugF("xf86VTSwitch()\n");
+    if(!(dispatchException & DE_TERMINATE))
+        assert(!seatd_libseat_controls_session());
 
 #ifdef XFreeXDGA
     if (!DGAVTSwitch())
diff --git a/hw/xfree86/common/xf86Init.c b/hw/xfree86/common/xf86Init.c
index 5695e71..0119d91 100644
--- a/hw/xfree86/common/xf86Init.c
+++ b/hw/xfree86/common/xf86Init.c
@@ -54,6 +54,7 @@
 #include "mi.h"
 #include "dbus-core.h"
 #include "systemd-logind.h"
+#include "seatd-libseat.h"
 
 #include "loaderProcs.h"
 
@@ -339,6 +340,7 @@ InitOutput(ScreenInfo * pScreenInfo, int argc, char **argv)
             DoShowOptions();
 
         dbus_core_init();
+        seatd_libseat_init();
         systemd_logind_init();
 
         /* Do a general bus probe.  This will be a PCI probe for x86 platforms */
@@ -847,6 +849,7 @@ ddxGiveUp(enum ExitCode error)
     if (xorgHWOpenConsole)
         xf86CloseConsole();
 
+    seatd_libseat_fini();
     systemd_logind_fini();
     dbus_core_fini();
 
diff --git a/hw/xfree86/common/xf86Xinput.c b/hw/xfree86/common/xf86Xinput.c
index 8cf3f62..2fe4c50 100644
--- a/hw/xfree86/common/xf86Xinput.c
+++ b/hw/xfree86/common/xf86Xinput.c
@@ -64,6 +64,7 @@
 #include "extinit.h"
 #include "loaderProcs.h"
 #include "systemd-logind.h"
+#include "seatd-libseat.h"
 
 #include "exevents.h"           /* AddInputDevice */
 #include "exglobals.h"
@@ -814,8 +815,10 @@ xf86DeleteInput(InputInfoPtr pInp, int flags)
 
     FreeInputAttributes(pInp->attrs);
 
-    if (pInp->flags & XI86_SERVER_FD)
+    if (pInp->flags & XI86_SERVER_FD) {
+        seatd_libseat_close_device(pInp);
         systemd_logind_release_fd(pInp->major, pInp->minor, pInp->fd);
+    }
 
     /* Remove the entry from the list. */
     if (pInp == xf86InputDevs)
@@ -940,9 +943,10 @@ xf86NewInputDevice(InputInfoPtr pInfo, DeviceIntPtr *pdev, BOOL enable)
     if (path && pInfo->major == 0 && pInfo->minor == 0)
         xf86stat(path, &pInfo->major, &pInfo->minor);
 
-    if (path && (drv->capabilities & XI86_DRV_CAP_SERVER_FD)){
+    if (path && (drv->capabilities & XI86_DRV_CAP_SERVER_FD)) {
         int fd = systemd_logind_take_fd(pInfo->major, pInfo->minor,
                                         path, &paused);
+        seatd_libseat_open_device(pInfo, &fd, &paused);
         if (fd != -1) {
             if (paused) {
                 /* Put on new_input_devices list for delayed probe */
diff --git a/hw/xfree86/os-support/linux/Makefile.am b/hw/xfree86/os-support/linux/Makefile.am
index 4392fe8..c58c063 100644
--- a/hw/xfree86/os-support/linux/Makefile.am
+++ b/hw/xfree86/os-support/linux/Makefile.am
@@ -19,6 +19,10 @@ ACPI_SRCS = $(srcdir)/../shared/pm_noop.c
 endif
 endif
 
+if SEATD_LIBSEAT
+LIBSEAT_SRCS = seatd-libseat.c
+endif
+
 if SYSTEMD_LOGIND
 LOGIND_SRCS = systemd-logind.c
 XORG_CFLAGS += $(DBUS_CFLAGS)
@@ -36,6 +40,7 @@ liblinux_la_SOURCES = linux.h lnx_init.c lnx_video.c \
 		     $(srcdir)/../shared/sigio.c \
                      $(ACPI_SRCS) \
                      $(APM_SRCS) \
+                     $(LIBSEAT_SRCS) \
                      $(LOGIND_SRCS) \
 		     $(AGP_SRCS)
 
diff --git a/hw/xfree86/os-support/linux/lnx_init.c b/hw/xfree86/os-support/linux/lnx_init.c
index 111b3b4..563dc26 100644
--- a/hw/xfree86/os-support/linux/lnx_init.c
+++ b/hw/xfree86/os-support/linux/lnx_init.c
@@ -36,6 +36,7 @@
 #include "xf86.h"
 #include "xf86Priv.h"
 #include "xf86_OSlib.h"
+#include "seatd-libseat.h"
 
 #include <sys/stat.h>
 #ifdef HAVE_SYS_SYSMACROS_H
@@ -213,6 +214,11 @@ xf86OpenConsole(void)
             FatalError("xf86OpenConsole: Cannot open virtual console"
                        " %d (%s)\n", xf86Info.vtno, strerror(errno));
 
+
+        /* If libseat is in control, it handles VT switching. */
+        if (seatd_libseat_controls_session())
+            return;
+
         /*
          * Linux doesn't switch to an active vt after the last close of a vt,
          * so we do this ourselves by remembering which is active now.
@@ -302,7 +308,7 @@ xf86CloseConsole(void)
     struct vt_stat vts;
     int ret;
 
-    if (xf86Info.ShareVTs) {
+    if (xf86Info.ShareVTs || seatd_libseat_controls_session()) {
         close(xf86Info.consoleFd);
         return;
     }
diff --git a/hw/xfree86/os-support/linux/lnx_platform.c b/hw/xfree86/os-support/linux/lnx_platform.c
index 8a6be97..db2dcdf 100644
--- a/hw/xfree86/os-support/linux/lnx_platform.c
+++ b/hw/xfree86/os-support/linux/lnx_platform.c
@@ -19,17 +19,23 @@
 
 #include "hotplug.h"
 #include "systemd-logind.h"
+#include "seatd-libseat.h"
 
 static Bool
 get_drm_info(struct OdevAttributes *attribs, char *path, int delayed_index)
 {
     drmVersionPtr v;
-    int fd;
+    int fd = -1;
     int err = 0;
     Bool paused, server_fd = FALSE;
 
     LogMessage(X_INFO, "Platform probe for %s\n", attribs->syspath);
 
+    fd = seatd_libseat_open_graphics(path);
+    if (fd != -1) {
+        attribs->fd = fd;
+        server_fd = TRUE;
+    }
     fd = systemd_logind_take_fd(attribs->major, attribs->minor, path, &paused);
     if (fd != -1) {
         if (paused) {
@@ -41,12 +47,14 @@ get_drm_info(struct OdevAttributes *attribs, char *path, int delayed_index)
         attribs->fd = fd;
         server_fd = TRUE;
     }
-
-    if (fd == -1)
+    if (fd == -1) {
+        // Try opening the path directly
         fd = open(path, O_RDWR | O_CLOEXEC, 0);
-
-    if (fd == -1)
-        return FALSE;
+        if (fd == -1) {
+            xf86Msg(X_ERROR, "cannot open %s\n", path);
+            return FALSE;
+        }
+    }
 
     /* for a delayed probe we've already added the device */
     if (delayed_index == -1) {
diff --git a/hw/xfree86/os-support/linux/seatd-libseat.c b/hw/xfree86/os-support/linux/seatd-libseat.c
new file mode 100644
index 0000000..b3120ed
--- /dev/null
+++ b/hw/xfree86/os-support/linux/seatd-libseat.c
@@ -0,0 +1,394 @@
+/*
+ * Copyright © 2022-2024 Mark Hindley, Ralph Ronnquist.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors: Mark Hindley <mark@hindley.org.uk>
+ *          Ralph Ronnquist <ralph.ronnquist@gmail.com>
+ */
+
+#ifdef HAVE_XORG_CONFIG_H
+#include <xorg-config.h>
+#endif
+
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <errno.h>
+#include <libseat.h>
+
+#include "os.h"
+#include "linux.h"
+#include "xf86.h"
+#include "xf86platformBus.h"
+#include "xf86Xinput.h"
+#include "xf86Priv.h"
+#include "globals.h"
+#include "seatd-libseat.h"
+
+// ============ libseat client adapter ======================
+
+struct libseat_info {
+    char *session;
+    Bool active;
+    Bool vt_active;
+    /*
+     * This pointer gets initialised to the actual libseat client instance
+     * provided by libseat_open_seat.
+     */
+    struct libseat *client;
+    int graphics_id;
+};
+static struct libseat_info seat_info;
+
+/*
+ * The seat has been enabled, and is now valid for use. Re-open all
+ * seat devices to ensure that they are operational, as existing fds
+ * may have had their functionality blocked or revoked.
+ */
+static void
+enable_seat(struct libseat *seat, void *userdata)
+{
+    InputInfoPtr pInfo;
+    (void) userdata;
+    LogMessage(X_INFO, "seatd_libseat enable\n");
+    seat_info.active = TRUE;
+    seat_info.vt_active = TRUE;
+
+    xf86VTEnter();
+    /* Reactivate all input devices */
+    for (pInfo = xf86InputDevs; pInfo; pInfo = pInfo->next)
+        if (pInfo->flags & XI86_SERVER_FD){
+            if (xf86CheckIntOption(pInfo->options, "libseat_id", -1) > 0){
+		int fd = -1, paused = FALSE;
+                seatd_libseat_open_device(pInfo, &fd, &paused);
+                xf86EnableInputDeviceForVTSwitch(pInfo);
+            }
+        }
+    xf86InputEnableVTProbe(); // Add any paused input devices
+    xf86platformVTProbe(); // Probe for outputs
+}
+
+/*
+ * The seat has been disabled. This event signals that the application
+ * is going to lose its seat access. The event *must* be acknowledged
+ * with libseat_disable_seat shortly after receiving this event.
+ *
+ * If the recepient fails to acknowledge the event in time, seat
+ * devices may be forcibly revoked by the seat provider.
+ */
+static void
+disable_seat(struct libseat *seat, void *userdata)
+{
+    (void) userdata;
+    LogMessage(X_INFO, "seatd_libseat disable\n");
+    xf86VTLeave();
+    seat_info.vt_active = FALSE;
+    if (libseat_disable_seat(seat)) {
+        LogMessage(X_ERROR, "seatd_libseat disable failed: %d\n", errno);
+    }
+}
+
+/*
+ * Callbacks for handling the libseat events.
+ */
+static struct
+libseat_seat_listener client_callbacks = {
+    .enable_seat = enable_seat,
+    .disable_seat = disable_seat,
+};
+
+/*
+ * Check libseat is initialised and active.
+ */
+static Bool
+libseat_active(void)
+{
+    if (!seat_info.client) {
+        LogMessageVerb(X_DEBUG, 5, "seatd_libseat not initialised!\n");
+        return FALSE;
+    }
+    if (!seat_info.active) {
+        LogMessage(X_DEBUG, "seatd_libseat not active\n");
+        return FALSE;
+    }
+    return TRUE;
+}
+
+/*
+ * Handle libseat events
+ */
+static int
+libseat_handle_events(int timeout)
+{
+    int ret;
+
+    while ((ret = libseat_dispatch(seat_info.client, timeout)) > 0)
+        LogMessage(X_INFO, "seatd_libseat handled %i events\n", ret);
+    if (ret == -1) {
+        LogMessage(X_ERROR, "libseat_dispatch() failed: %s\n", strerror(errno));
+        return -1;
+    }
+    return ret;
+}
+
+static void
+event_handler(int fd, int ready, void *data)
+{
+    LogMessage(X_INFO, "seatd_libseat event handler\n");
+    libseat_handle_events(0);
+}
+
+/*
+ * Handle libseat logging.
+ */
+static void
+log_libseat(enum libseat_log_level level, const char *fmt, va_list args)
+{
+    MessageType xmt;
+    size_t xfmt_size = strlen(fmt) + 2;
+    char *xfmt;
+
+    xfmt = malloc(xfmt_size);
+    if (xfmt == NULL)
+        return;
+    snprintf(xfmt, xfmt_size, "%s\n", fmt);
+
+    switch (level) {
+    case LIBSEAT_LOG_LEVEL_INFO:
+        xmt = X_INFO;
+        break;
+    case LIBSEAT_LOG_LEVEL_ERROR:
+        xmt = X_ERROR;
+        break;
+    default:
+        xmt = X_DEBUG;
+    }
+    LogVMessageVerb(xmt, 0, xfmt, args);
+
+    free(xfmt);
+}
+
+// ============== seatd-libseat.h API functions =============
+
+/*
+ * Initialise the libseat client.
+ *
+ * Returns:
+ *   0 if all ok
+ *   1 if not possible
+ *   -EPERM (-1) if it was already initialised
+ *   -EPIPE (-32) if the seat opening failed.
+ */
+int
+seatd_libseat_init(void)
+{
+    if (!ServerIsNotSeat0() && xf86HasTTYs() && linux_parse_vt_settings(TRUE) && !linux_get_keeptty()) {
+        LogMessage(X_INFO,
+            "seat-libseat: libseat integration requires -keeptty and "
+            "-keeptty was not provided, disabling libseat integration\n");
+        return 1;
+    }
+
+    libseat_set_log_level(LIBSEAT_LOG_LEVEL_DEBUG);
+    libseat_set_log_handler(log_libseat);
+    LogMessage(X_INFO, "seatd_libseat init\n");
+    if (libseat_active()) {
+        LogMessage(X_ERROR, "seatd_libseat already initialised\n");
+        return -EPERM;
+    }
+    seat_info.graphics_id = -1;
+    seat_info.client = libseat_open_seat(&client_callbacks, NULL);
+    if (!seat_info.client) {
+        LogMessage(X_ERROR, "Cannot set up seatd_libseat client\n");
+        return -EPIPE;
+    }
+    SetNotifyFd(libseat_get_fd(seat_info.client), event_handler, X_NOTIFY_READ, NULL);
+
+    if (libseat_handle_events(100) < 0) {
+        libseat_close_seat(seat_info.client);
+        return -EPIPE;
+    }
+    LogMessage(X_INFO, "seatd_libseat client activated\n");
+    return 0;
+}
+
+/*
+ * Shutdown the libseat client.
+ */
+void
+seatd_libseat_fini(void)
+{
+    if (seat_info.client) {
+        LogMessage(X_INFO, "seatd_libseat finish\n");
+        libseat_close_seat(seat_info.client);
+    }
+    seat_info.graphics_id = -1;
+    seat_info.active = FALSE;
+    seat_info.client = NULL;
+}
+
+/*
+ * Open the graphics device
+ *
+ * Return
+ *   file descriptor (>=0) if all is ok.
+ *   -EPERM (-1) if the libseat client is not activated
+ *   -EAGAIN (-11) if the VT is not active
+ *   -errno from libseat_open_device if device access failed
+ */
+int
+seatd_libseat_open_graphics(const char *path)
+{
+    int fd, id;
+
+    if (!libseat_active()) {
+        return -EPERM;
+    }
+    LogMessage(X_INFO, "seatd_libseat try open graphics %s\n", path);
+    if ((id = libseat_open_device(seat_info.client, path, &fd)) == -1) {
+        fd = -errno;
+        LogMessage(X_ERROR, "seatd_libseat open graphics %s (%d) failed: %d\n",
+                   path, id, fd);
+    }
+    else {
+        LogMessage(X_INFO, "seatd_libseat opened graphics: %s (%d:%d)\n", path,
+                   id, fd);
+    }
+    seat_info.graphics_id = id;
+    return fd;
+}
+
+/*
+ * Find duplicate devices with same major:minor number and assigned
+ * "libseat_id" and, if any, return its file descriptor.
+ */
+static int
+check_duplicate_device(int maj, int min) {
+
+    InputInfoPtr pInfo;
+
+    for (pInfo = xf86InputDevs; pInfo; pInfo = pInfo->next) {
+        if (pInfo->major == maj && pInfo->minor == min &&
+            xf86CheckIntOption(pInfo->options, "libseat_id", -1) >= 0) {
+            return pInfo->fd;
+        }
+    }
+    return -1;
+}
+
+/*
+ * Open an input device.
+ *
+ * The function sets the p->options "libseat_id" for the device when
+ * successful.
+ */
+void
+seatd_libseat_open_device(InputInfoPtr p, int *pfd, Bool *paused)
+{
+    int id, fd;
+    char *path = xf86CheckStrOption(p->options, "Device", NULL);
+
+    if (!libseat_active()) {
+        return;
+    }
+    if (!seat_info.vt_active) {
+	*pfd = -2; /* Invalid, but not -1. See xf86NewInputDevice() */
+	*paused = TRUE;
+        LogMessage(X_INFO, "seatd_libseat paused %s\n", path);
+	return;
+    }
+    fd = check_duplicate_device(p->major,p->minor);
+    if (fd < 0) {
+        LogMessage(X_INFO, "seatd_libseat try open %s\n", path);
+        if ((id = libseat_open_device(seat_info.client, path, &fd)) == -1) {
+            fd = -errno;
+            LogMessage(X_ERROR, "seatd_libseat open %s (%d) failed: %d\n",
+                       path, id, fd);
+            return;
+        }
+    }
+    else {
+        LogMessage(X_INFO, "seatd_libseat reuse %d for %s\n", fd, path);
+    }
+    p->flags |= XI86_SERVER_FD;
+    p->fd = fd;
+    p->options = xf86ReplaceIntOption(p->options, "fd", fd);
+    p->options = xf86ReplaceIntOption(p->options, "libseat_id", id);
+    LogMessage(X_INFO, "seatd_libseat opened %s (%d:%d)\n", path, id, fd);
+}
+
+/*
+ * Release an input device.
+ */
+void
+seatd_libseat_close_device(InputInfoPtr p)
+{
+    char *path = xf86CheckStrOption(p->options, "Device", NULL);
+    int fd = xf86CheckIntOption(p->options, "fd", -1);
+    int id = xf86CheckIntOption(p->options, "libseat_id", -1);
+
+    if (!libseat_active())
+        return;
+    LogMessage(X_INFO, "seatd_libseat try close %s (%d:%d)\n", path, id, fd);
+    if (fd < 0) {
+        LogMessage(X_ERROR, "seatd_libseat device not open (%s)\n", path);
+        return;
+    }
+    if (id < 0) {
+        LogMessage(X_ERROR, "seatd_libseat no libseat ID\n");
+        return;
+    }
+    if (libseat_close_device(seat_info.client, id)) {
+        LogMessage(X_ERROR, "seatd_libseat close failed %d\n", -errno);
+    }
+    else {
+        close(fd);
+        p->fd = -1;
+        p->options = xf86ReplaceIntOption(p->options, "fd", -1);
+    }
+}
+
+/*
+ * Libseat controls session
+ */
+
+Bool
+seatd_libseat_controls_session(void){
+    return libseat_active();
+}
+
+/*
+ * Switch VT
+ */
+int
+seatd_libseat_switch_session(int session)
+{
+    int ret=0;
+
+    LogMessage(X_INFO, "seatd_libseat switch VT %d\n", session);
+    if ((ret = libseat_switch_session(seat_info.client, session)) < 0) {
+        LogMessage(X_ERROR, "seatd_libseat switch VT failed with %d\n", -errno);
+        goto ret;
+    }
+ ret:
+    return ret;
+}
diff --git a/hw/xfree86/os-support/shared/VTsw_usl.c b/hw/xfree86/os-support/shared/VTsw_usl.c
index 6440261..967669f 100644
--- a/hw/xfree86/os-support/shared/VTsw_usl.c
+++ b/hw/xfree86/os-support/shared/VTsw_usl.c
@@ -31,6 +31,8 @@
 #include "xf86Priv.h"
 #include "xf86_OSlib.h"
 
+#include "seatd-libseat.h"
+
 /*
  * Handle the VT-switching interface for OSs that use USL-style ioctl()s
  * (the sysv, sco, and linux subdirs).
@@ -58,6 +60,8 @@ Bool
 xf86VTSwitchAway(void)
 {
     xf86Info.vtRequestsPending = FALSE;
+    if (seatd_libseat_controls_session())
+        return TRUE;
     if (ioctl(xf86Info.consoleFd, VT_RELDISP, 1) < 0)
         return FALSE;
     else
@@ -68,6 +72,8 @@ Bool
 xf86VTSwitchTo(void)
 {
     xf86Info.vtRequestsPending = FALSE;
+    if (seatd_libseat_controls_session())
+        return TRUE;
     if (ioctl(xf86Info.consoleFd, VT_RELDISP, VT_ACKACQ) < 0)
         return FALSE;
     else
diff --git a/include/Makefile.am b/include/Makefile.am
index 25b3a90..d66103f 100644
--- a/include/Makefile.am
+++ b/include/Makefile.am
@@ -75,6 +75,7 @@ EXTRA_DIST = 	\
 	eventconvert.h eventstr.h inpututils.h \
 	probes.h \
 	protocol-versions.h \
+	seatd-libseat.h \
 	swaprep.h \
 	swapreq.h \
 	systemd-logind.h \
diff --git a/include/dix-config.h.in b/include/dix-config.h.in
index 382d706..5d29337 100644
--- a/include/dix-config.h.in
+++ b/include/dix-config.h.in
@@ -415,6 +415,9 @@
 /* Support HAL for hotplug */
 #undef CONFIG_HAL
 
+/* Enable libseat integration */
+#undef SEATD_LIBSEAT
+
 /* Enable systemd-logind integration */
 #undef SYSTEMD_LOGIND 1
 
diff --git a/include/meson.build b/include/meson.build
index 591bc25..bfcf3a0 100644
--- a/include/meson.build
+++ b/include/meson.build
@@ -96,6 +96,7 @@ conf_data.set('CONFIG_UDEV', build_udev ? '1' : false)
 conf_data.set('CONFIG_UDEV_KMS', build_udev_kms ? '1' : false)
 conf_data.set('HAVE_DBUS', build_dbus ? '1' : false)
 conf_data.set('CONFIG_HAL', build_hal ? '1' : false)
+conf_data.set('SEATD_LIBSEAT', build_seatd_libseat ? '1' : false)
 conf_data.set('SYSTEMD_LOGIND', build_systemd_logind ? '1' : false)
 conf_data.set('NEED_DBUS', build_systemd_logind or build_hal ? '1' : false)
 conf_data.set('CONFIG_WSCONS', host_machine.system() == 'openbsd' ? '1' : false)
diff --git a/include/seatd-libseat.h b/include/seatd-libseat.h
new file mode 100644
index 0000000..cbc0d2a
--- /dev/null
+++ b/include/seatd-libseat.h
@@ -0,0 +1,49 @@
+/*
+ * Copyright © 2022-2024 Mark Hindley, Ralph Ronnquist.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors: Mark Hindley <mark@hindley.org.uk>
+ *          Ralph Ronnquist <ralph.ronnquist@gmail.com>
+ */
+
+#ifndef SEATD_LIBSEAT_H
+#define SEATD_LIBSEAT_H
+
+#ifdef SEATD_LIBSEAT
+#include <xf86Xinput.h>
+extern int seatd_libseat_init(void);
+extern void seatd_libseat_fini(void);
+extern int seatd_libseat_open_graphics(const char *path);
+extern void seatd_libseat_open_device(InputInfoPtr p,int *fd,Bool *paus);
+extern void seatd_libseat_close_device(InputInfoPtr p);
+extern int seatd_libseat_switch_session(int session);
+extern Bool seatd_libseat_controls_session(void);
+#else
+#define seatd_libseat_init()
+#define seatd_libseat_fini()
+#define seatd_libseat_open_graphics(path) -1
+#define seatd_libseat_open_device(p,x,y)
+#define seatd_libseat_close_device(p)
+#define seatd_libseat_switch_session(int) -1
+#define seatd_libseat_controls_session() FALSE
+#endif
+
+#endif
diff --git a/meson.build b/meson.build
index 9e54a0a..1fb2eac 100644
--- a/meson.build
+++ b/meson.build
@@ -172,8 +172,14 @@ if ['windows',  'darwin', 'cygwin'].contains(host_machine.system())
     hal_option = 'false'
 endif
 
+if get_option('seatd_libseat') == 'auto'
+    build_seatd_libseat = build_udev_kms
+else
+    build_seatd_libseat = get_option('seatd_libseat') == 'true'
+endif
+
 if get_option('systemd_logind') == 'auto'
-    build_systemd_logind = build_udev_kms and dbus_dep.found()
+    build_systemd_logind = build_udev_kms and dbus_dep.found() and not build_seatd_libseat
 else
     build_systemd_logind = get_option('systemd_logind') == 'true'
 endif
diff --git a/meson_options.txt b/meson_options.txt
index b12d521..a51fab4 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -67,6 +67,8 @@ option('hal', type: 'combo', choices: ['true', 'false', 'auto'], value: 'auto',
        description: 'Enable HAL integration')
 option('systemd_logind', type: 'combo', choices: ['true', 'false', 'auto'], value: 'auto',
        description: 'Enable systemd-logind integration')
+option('seatd_libseat', type: 'combo', choices: ['true', 'false', 'auto'], value: 'auto',
+       description: 'Enable seatd-libseat integration')
 option('vgahw', type: 'combo', choices: ['true', 'false', 'auto'], value: 'auto',
        description: 'Xorg VGA access module')
 option('dpms', type: 'boolean', value: true,
